### PR TITLE
refactor spikeGLX

### LINKED ISSUE #290: [Feature]: reorganize SpikeGLX interfaces
### What would you like to see added to NeuroConv?

Right now we have:

ap -> `SpikeGLXRecordingInterface`
lf -> `SpikeGLXLFPInterface`
nidq -> `SpikeGLXNIDQInterface`

This scheme is useful because each interface uses the appropriate extractor in the standard way.

The main problem with this layout is that it is too difficult for a user to create a conversion that includes all 3. Only 1 of them is the "recording" interface, but from a NeuroConv perspective they really all should be, because they are all data produced by a SpikeGLX recording. The ap is the stream that generally goes into spike sorting pipelines, but that doesn't make it more of a recording than lf or nidq. It often makes sense to differentiate between a recording and an LFP stream because often LFP is derived from the raw unfiltered stream, but that is not the case here. In this case, I think having a "recording" that does not include all three is misleading.

I would like to propose the following change:

ap -> `SpikeGLXAPInterface`, which creates an ElectricalSeriesAP (renaming the current SpikeGLXRecordingInterface)
lf -> `SpikeGLXLFInterface`, which creates an ElectricalSeriesLF (renaming the current SpikeGLXLFPInterface)
nidq -> `SpikeGLXNIDQInterface` (this is unchanged)

all three together -> `SpikeGLXRecordingInterface`

The `SpikeGLXRecordingInterface` would take as input the entire directory of the spikeglx recording and flags for what streams to include, defaulting to including all of them. It would automatically infer the paths of the ap, lf, and nidq streams and would load the data and metadata and do the conversion very similarly to how an NWBConverter works. If there is something non-standard, like a different folder structure for the spikeglx data, a user can always go down to the individual interfaces. The conversion gallery would continue to show the `SpikeGLXRecordingInterface` in its default configuration, which would convert all three streams.

### Is your feature request related to a problem?

_No response_

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/catalystneuro/neuroconv/blob/main/.github/CODE_OF_CONDUCT.rst)
- [X] Have you ensured this bug was not already [reported](https://github.com/catalystneuro/neuroconv/issues)?

### PR DESCRIPTION
**Pull request by bendichter**
## Motivation
fix #290 

SpikeGLXRecordingInterface can now be used for ap and lf. It assigns es_key based on the filepath

### CODE REVIEW

**Comment by CodyCBakerPhD** [comment]
OK so this approach still allows manual override if necessary, just shifting it from the `conversion_options` to the `__init__` to that the reference between `get_metadata` and `run_conversion` are in agreement

Works for me, let me know when tests are passing for a final look

One question is how to handle the conversion gallery, since this is the first time we've deprecated a popular interface...

After the next release that includes this PR, anyone running the current gallery will get a deprecation warning - not the best thing for a user to see upon their first attempts using our package

Then again, if we update the gallery now and they try it with a previous version of NeuroConv, the interface won't even exist so it'll be an import error...

Only thing I can think of is versioning the gallery/docs to make it clear what's on dev vs. most recent released versions

**Comment by bendichter** [comment]
I realized that we should map to names of ElectricalSeries that allow for multi probes. For right now I am simply using the stream_id, e.g. the ElectricalSeries for ap 0 and lf 0 are called "imec0.ap" and the "imec0.lf". This solves the problem but may not be the best choice for readability. I suppose I would be open to changing to "ElectricalSeries_imec0.ap" or similar to keep with best practices

**Comment by bendichter** [comment]
@CodyCBakerPhD I moved so much logic up into BaseRecordingExtractorInterface that I don't even need different types for processing the AP and LF streams, so I removed `SpikeGLXInterface` altogether and just made it a copy of `SpikeGLXRecordingInterface` for backward compatibility.

File: src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py [hunk 1]
```diff
@@ -38,18 +35,17 @@ def add_recording_extractor_properties(recording_extractor) -> None:
...
 
 class SpikeGLXRecordingInterface(BaseRecordingExtractorInterface):
-    """Primary data interface class for converting the high-pass (ap) SpikeGLX format."""
+    ExtractorName = "SpikeGLXRecordingExtractor"
 
     @classmethod
     def get_source_schema(cls) -> dict:
         source_schema = get_schema_from_method_signature(class_method=cls.__init__, exclude=["x_pitch", "y_pitch"])
-        source_schema["properties"]["file_path"]["description"] = "Path to SpikeGLX file."
+        source_schema["properties"]["file_path"]["description"] = "Path to SpikeGLX ap.in or lf.bin file."
```
Comment (bendichter) [inline src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py:43]:
```suggestion
        source_schema["properties"]["file_path"]["description"] = "Path to SpikeGLX ap.bin or lf.bin file."
```

**Comment by bendichter** [comment]
@CodyCBakerPhD this PR does 3 things:
1. It deletes some automated BlackRock behavior that put certain ns files in processing rather than acquisition. I don't think the automated decision making was necessarily right and the user can use the "write_as" arg to control that manually.
2. It makes `es_key` an attribute of the `BaseRecordingExtractorInterface`. It defaults to "ElectricalSeries", which can be overridden as an arg in the `__init__`. The es_key arg is no longer available for `run_conversion`.
3. `SpikeGLXLFPInterface` has been deleted. This can now be done entirely with `SpikeGLXRecordingExtractor`, which infers whether it is an ap or lf stream based on the input path.


I know this is a lot for 1 PR. I was just trying to accomplish 3 but then 1 and 2 ended up being necessary. I can separate these out into individual PRs if that will make them easier for you to review.

**Comment by bendichter** [comment]
nvm looks like a lot of bugs just cropped up

**Comment by bendichter** [comment]
depends on #318

**Comment by bendichter** [comment]
@CodyCBakerPhD thanks for helping me push these through. This is the next big one.

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,9 @@
 * The `OpenEphysRecordingInterface` now redirects to legacy or binary interface depending on the file format.
   It raises NotImplementedError until the interface for legacy format is added. [PR #296](https://github.com/catalystneuro/neuroconv/pull/296)
 * Added `PlexonSortingInterface` to support plexon spiking data. [PR #316](https://github.com/catalystneuro/neuroconv/pull/316)
+* Move `es_key` to the `__init__` of the `BaseRecordingExtractorInterface` class. It is no longer possible to use
+  this argument in the `run_conversion` method. [PR #318](https://github.com/catalystneuro/neuroconv/pull/318)
```
Comment (bendichter) [inline CHANGELOG.md:17]:
```suggestion
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -16,6 +16,9 @@
   It raises NotImplementedError until the interface for legacy format is added. [PR #296](https://github.com/catalystneuro/neuroconv/pull/296)
 * Added the `OpenEphysLegacyRecordingInterface` to support Open Ephys legacy format (`.continuous` files). [PR #295](https://github.com/catalystneuro/neuroconv/pull/295)
 * Added `PlexonSortingInterface` to support plexon spiking data. [PR #316](https://github.com/catalystneuro/neuroconv/pull/316)
+* Move `es_key` to the `__init__` of the `BaseRecordingExtractorInterface` class. It is no longer possible to use
+  this argument in the `run_conversion` method. [PR #318](https://github.com/catalystneuro/neuroconv/pull/318)
```
Comment (bendichter) [inline CHANGELOG.md:20]:
```suggestion
```

File: src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglx_utils.py [hunk 4]
```diff
@@ -62,3 +63,34 @@ def fetch_stream_id_for_spikelgx_file(file_path: FilePathType) -> str:
...
+
+    if "imDatBsc_pn" in meta:
+        metadata_dict.update(connected_base_station_part_number=meta["imDatBsc_pn"])
+
+    description_string = "no description"
+    if metadata_dict:
+        description_string = json.dumps(metadata_dict)
+    device_metadata = dict(name="Neuropixel-Imec", description=description_string, manufacturer="Imec")
+
+    return device_metadata
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglx_utils.py:96]: Did you mean to delete this as a method in the main class? https://github.com/catalystneuro/neuroconv/blob/51c254bf28952f8d4da24070bafab1d7af552433/src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py#L118-L148

File: src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py [hunk 5]
```diff
@@ -146,7 +148,9 @@ def get_device_metadata(self) -> dict:
         return device
 
 
+# include for backwards compatibility
 class SpikeGLXLFPInterface(SpikeGLXRecordingInterface):
+    pass
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py:153]: Did you want to eventually deprecate it altogether? If so just need a warning here

File: src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglx_utils.py (hunk 4 above)
Comment (bendichter) [reply src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglx_utils.py:96]: yes, I did. I think it got back in there with merges. I'll remove it again

**Comment by CodyCBakerPhD** [comment]
Two small comments otherwise LGTM

**Review by CodyCBakerPhD — DISMISSED** [review]

**Review by CodyCBakerPhD — APPROVED** [review]

**Pull request merged**